### PR TITLE
[MPS] Change exponential to return values in  `(0, inf)` instead of `[0, inf)` range

### LINKED ISSUE #192577: [MPS] multinomial(p, 1) intermittently returns zero-probability indices — exponential_() emits exact -0.0, Gumbel fast path turns it into a NaN that argmax selects
### 🐛 Describe the bug

**Expected:** `torch.multinomial(p, 1)` never returns an index `i` with `p[i] == 0`.

**Actual:** on MPS it does so intermittently — ~1 per ~220 draws for a 152k-category distribution whose tail underflows to exact zeros (i.e., any temperature-sharpened softmax, any top-k/top-p filtered distribution, any masked policy). CPU: zero occurrences.

**Repro (~60 s, no model):**

```python
import torch
V = 151936
logits = torch.randn(V) * 2.0
logits[12345] += 200.0                    # tails underflow to exact 0.0
p = torch.softmax(logits, -1).to("mps")
bad = sum(float(p[int(torch.multinomial(p, 1))]) == 0.0 for _ in range(20_000))
print(bad)   # M1 Pro, macOS 15.7.7, torch 2.13.0: 90.  CPU: 0.
```

**Root cause** (each step measured; full forensic dataset available — ~477k individually verified draws, 695 corruption events, saved tensors):

1. MPS `exponential_(1)` emits exact zeros at ~3.3e-8/element (census: 10 zeros in 3.04e8 samples; CPU: 0 in the same count). Captured values are `-0.0` (signbit set), the fingerprint of a negated log at a boundary uniform. `Distributions.cpp` already documents this hazard class — *"on CUDA side there is a protection to avoid 0s"* — MPS has no such protection.
2. The n=1 Gumbel fast path (`q = exponential_; q = p/q; argmax`, `aten/src/ATen/native/Distributions.cpp` ~594–628) computes `0 / -0 = NaN` when the RNG zero lands on a zero-probability entry (one element per affected draw; over a 152k vocab that's ~0.5% of draws).
3. MPS argmax propagates NaN as the maximum, so the returned "sample" is the position of the RNG zero — an index unrelated to the distribution. (When the zero lands on a positive-probability entry, `p / -0 = -inf` loses argmax and the defect is invisible.)
4. Fully deterministic given RNG state: restoring the state reproduces the identical corrupt index (45/45 replays); `torch.mps.synchronize()` has no effect; reproduces at rest. Not a race.

**Impact:** silent wrong-token emission in any sampled decoding on Apple Silicon (`do_sample` / top-k / top-p; greedy and beam are unaffected) at ~1 per few hundred sampled tokens. Discovered via an LLM agent that began investigating objects that didn't exist — the model's own distribution wanted the correct token at 99.9% while the sampler emitted a token with recorded probability exactly 0.0; traces available.

**Suggested fix:** protect the exponential output *before* the division — e.g. clamp the MPS `exponential_` kernel's output to a positive minimum, mirroring the CUDA protection. (Post-division guards don't work: `clamp` propagates the NaN, and `nan_to_num` would also mask genuine `p` NaNs that the current `_assert_async` checks are designed to catch.)

**Workaround for users:** draw on CPU — `torch.multinomial(p.cpu(), 1)`.

Relates to #136623 (same op, different failure surface: that issue is out-of-domain indices from the multi-sample kernel; this one is in-domain zero-probability indices from the single-sample fast path).

### Versions

<details>

```
Collecting environment information...
PyTorch version: 2.13.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.7.7 (arm64)
GCC version: Could not collect
Clang version: 16.0.0 (clang-1600.0.26.4)
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.4 (v3.12.4:8e8a4baf65, Jun  6 2024, 17:33:18) [Clang 13.0.0 (clang-1300.0.29.30)] (64-bit runtime)
Python platform: macOS-15.7.7-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Apple M1 Pro

Versions of relevant libraries:
```

</details>

cc @pbelevich @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20 @albanD

### PR DESCRIPTION
**Pull request by Srinandha-Murugesan**
Fixes #192577: MPS issue where `torch.multinomial(p, 1)` can select an index with a probability of zero.

Which is fine per note in the documentation: https://docs.pytorch.org/docs/2.13/generated/torch.Tensor.exponential_.html

### CODE REVIEW

File: test/test_mps.py
Comment (Isalia20) [inline test/test_mps.py:1]: This seems like an important test to have for other devices as well so please move the test to OpInfo tests so if similar thing occurs on CPU/CUDA we catch it. `test_mps.py` should be for tests that exercise Metal kernels for example or something MPS specific.

File: aten/src/ATen/native/mps/kernels/Distributions.metal [hunk 1]
```diff
@@ -98,8 +98,12 @@ kernel void exponential(
     // Only `u = 1` is the failure mode (`log(0)`); `u = 0` gives `log(1) = 0`.
     float u = ::metal::min(
         c10::metal::detail::uint32_to_uniform_float(raw[i]), 1.0f - eps);
-    output[base + i] =
-        static_cast<T>(-::metal::precise::log(1.0f - u) / lambda);
+
+    // Negation of `log(1)` can produce `-0.0` for `u = 0` and
+    // can lead to `0 / -0.0 = NaN` in n=1 Gumbel fast path.
+    float val = -::metal::precise::log(1.0f - u) / lambda;
+    val = ::metal::max(val, FLT_MIN);
```
Comment (Isalia20) [inline aten/src/ATen/native/mps/kernels/Distributions.metal]: This will underflow to 0.0 for fp16, please include fp32/bf16/fp16 in tests to catch such things

**Review by Isalia20 — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/mps/kernels/Distributions.metal [hunk 2]
```diff
@@ -98,8 +98,12 @@ kernel void exponential(
...
+    output[base + i] = static_cast<T>(val);
```
Comment (Isalia20) [inline aten/src/ATen/native/mps/kernels/Distributions.metal]: i meant when you cast the fp16 value here

**Comment by Srinandha-Murugesan** [comment]
@Isalia20 I fixed the clamp to work for the different types and moved the general draw regression test in test_ops.py. My change passed for all three types in MPS.

While validating it on CPU, I found that the same issue happened in CPU fp16. I looked into it a bit and the CPU multinomial implementation doesn't use exponential kernel so this issue isn't related to my changes. I'm also not sure whether the CPU behavior is something that should be addressed as part of this PR or a separate issue/PR.

File: test/test_mps.py
Comment (malfet) [reply test/test_mps.py:1]: I don't think it can be moved to opinfo, but I suspect something like that must exist elsewhere

File: aten/src/ATen/native/mps/kernels/Distributions.metal [hunk 3]
```diff
@@ -95,9 +95,10 @@ kernel void exponential(
   float lambda = params.x;
   uint count = min(4u, numel - base);
   for (uint i = 0; i < count; ++i) {
-    // Only `u = 1` is the failure mode (`log(0)`); `u = 0` gives `log(1) = 0`.
-    float u = ::metal::min(
-        c10::metal::detail::uint32_to_uniform_float(raw[i]), 1.0f - eps);
+    // Clamp `u` away from 0 to prevent `1.0f - u` from rounding to 1.0f
+    // and producing `-0.0` from `-log(1.0f - u)`.
+    float u = ::metal::clamp(
```
Comment (malfet) [inline aten/src/ATen/native/mps/kernels/Distributions.metal:100]: CUDA/CPU do eps/2 as lower bound, if I'm reading transformer helper correctly...

File: aten/src/ATen/native/mps/kernels/Distributions.metal (hunk 2 above)
Comment (robertomeroni) [reply aten/src/ATen/native/mps/kernels/Distributions.metal]: I hit the same `-0.0` bug from the multinomial side and measured this on an M3 Pro (macOS 26.6) with the `clamp` fix applied, in case it saves you a round trip. Output of my runs:

```
    -log(1 - eps) as f32 = 1.192093038e-07
      cast -> float16    = 1.192092896e-07   ok
      cast -> bfloat16   = 1.192092896e-07   ok
      cast -> float32    = 1.192093038e-07   ok
```

So the clamped value itself survives the cast in all three dtypes -- f16's smallest subnormal is
5.96e-08, just under it. What does not survive is the division, since `lambda` is applied after the
log and before the cast:

```
    lambda=1    f32 1.192093038e-07 -> f16 1.192092896e-07  ok
    lambda=2    f32 5.960465188e-08 -> f16 5.960464478e-08  ok
    lambda=4    f32 2.980232594e-08 -> f16 5.960464478e-08  ok
    lambda=8    f32 1.490116297e-08 -> f16 0.0              ZERO
    lambda=100  f32 1.192093091e-09 -> f16 0.0              ZERO
```

Measured end to end with the fix in place, `exponential_(lambda)` in float16:

```
    MPS, zeros per 1e8 samples:  lambda=1 -> 0,  4 -> 14,  8 -> 19,  100 -> 265
    CPU, zeros per 2e7 samples:  lambda=1 -> 0,  4 ->  2,  8 ->  3,  100 ->  70
```

Two things worth separating: these are `+0.0` (sign bit clear) from underflow, not the `-0.0` this PR
fixes, and CPU produces them at the same per-element rate -- so it looks pre-existing and backend-
independent rather than something this change introduces or should have to solve. `signbit` was clear
on all 1e8 MPS samples with the clamp in, at every lambda I tried.

File: aten/src/ATen/native/mps/kernels/Distributions.metal (hunk 3 above)
Comment (robertomeroni) [reply aten/src/ATen/native/mps/kernels/Distributions.metal:100]: confirmed -- TransformationHelper.h:137 floors the CUDA output at eps/2, and clamping u to eps/2 here
reproduces that exactly (eps gives 2x it). Neat side effect: eps/2 is 2^-24, f16's smallest subnormal,
so it is also the smallest bound that survives the cast. 

note the CPU branch is log1p(-val) with no lower clamp, so it is really just CUDA that does eps/2

File: test/test_mps.py
Comment (robertomeroni) [reply test/test_mps.py:1]: `test_exponential_no_zero` at test/test_torch.py:2253 is the one -- device-generic already, asserts
`x.min() > 0`, but gated `@onlyCUDA`. Un-gating it for MPS needs `allow_mps=True` on the
`TestTorchDeviceType` instantiation, which pulls in ~750 other MPS tests, so probably its own PR

**Comment by robertomeroni** [comment]
@Srinandha-Murugesan 
>While validating it on CPU, I found that the same issue happened in CPU fp16. I looked into it a bit and the CPU multinomial implementation doesn't use exponential kernel so this issue isn't related to my changes. I'm also not sure whether the CPU behavior is something that should be addressed as part of this PR or a separate issue/PR.

the cpu_float16 failure is pre-existing, it reproduces on a stock 2.9.1 wheel. CPU does use the same Gumbel path (Distributions.cpp:614), so it isn't multinomial: exponential_ samples in double and static_casts to
scalar_t (cpu/DistributionTemplates.h:341), so anything below fp16's 2^-24 becomes exactly 0, p/q is
0/0 = NaN, and argmax returns that index. Same eps/2 bound as the other thread -- the CUDA branch of
TransformationHelper.h:137 floors there and is immune. Dropping float16 from the CPU arm would unblock
you; happy to file the underflow separately

File: test/test_mps.py [hunk 4]
```diff
@@ -1017,6 +1017,48 @@ def test_exp(self, device="mps", dtype=torch.float):
...
+                torch.mps.set_rng_state(bad_rng_state)
+
+                idx = torch.multinomial(p, 1)
+
+                self.assertGreater(
+                    p[idx].item(),
+                    0.0,
+                    f"torch.multinomial selected a zero-probability index on MPS for dtype {dtype}",
+                )
+
```
Comment (malfet) [inline test/test_mps.py]:
```suggestion
```

File: test/test_ops.py [hunk 5]
```diff
@@ -24,6 +24,7 @@
 from torch.testing._internal.common_cuda import with_tf32_off
 from torch.testing._internal.common_device_type import (
     deviceCountAtLeast,
+    dtypes,
```
Comment (malfet) [inline test/test_ops.py]:
```suggestion
```

File: test/test_ops.py [hunk 6]
```diff
@@ -338,6 +339,28 @@ def get_opoverloadpacket_from_dispatch(kernel):
...
+        num_draws = 20_000
+        for _ in range(num_draws):
+            idx = torch.multinomial(p, 1)
+            self.assertGreater(
+                p[idx].item(),
+                0.0,
+                f"torch.multinomial selected a zero-probability index "
+                f"on {device} ({dtype})",
+            )
+
```
Comment (malfet) [inline test/test_ops.py]:
```suggestion
```

**Comment by Srinandha-Murugesan** [comment]
I just made it eps/2.

**Comment by Srinandha-Murugesan** [comment]
> @Srinandha-Murugesan
> 
> > While validating it on CPU, I found that the same issue happened in CPU fp16. I looked into it a bit and the CPU multinomial implementation doesn't use exponential kernel so this issue isn't related to my changes. I'm also not sure whether the CPU behavior is something that should be addressed as part of this PR or a separate issue/PR.
> 
> the cpu_float16 failure is pre-existing, it reproduces on a stock 2.9.1 wheel. CPU does use the same Gumbel path (Distributions.cpp:614), so it isn't multinomial: exponential_ samples in double and static_casts to scalar_t (cpu/DistributionTemplates.h:341), so anything below fp16's 2^-24 becomes exactly 0, p/q is 0/0 = NaN, and argmax returns that index. Same eps/2 bound as the other thread -- the CUDA branch of TransformationHelper.h:137 floors there and is immune. Dropping float16 from the CPU arm would unblock you; happy to file the underflow separately

@robertomeroni yes thank you, that would be great.

**Review by malfet — APPROVED** [review]
